src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -1660,92 +1660,92 @@ static int client_set_lease_timeouts(sd_dhcp_client *client) {
-                if (t2_timeout <= t1_timeout) {
+                if (client->t2_time <= client->t1_time) {
                         /* the computed T1 would be invalid, so discard T2 */
-                        t2_timeout = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
+                        client->t2_time = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
                         client->lease->t2 = (client->lease->lifetime * 7) / 8;
                 }
         } else if (client->lease->t1 > 0 && client->lease->t1 < client->lease->lifetime) {
                 /* only T1 is given, and it is valid */
-                t1_timeout = client_compute_timeout(client, client->lease->t1, 1);
-                t2_timeout = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
+                client->t1_time = client_compute_timeout(client, client->lease->t1, 1);
+                client->t2_time = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
                 client->lease->t2 = (client->lease->lifetime * 7) / 8;
-                if (t2_timeout <= t1_timeout) {
+                if (client->t2_time <= client->t1_time) {
                         /* the computed T2 would be invalid, so discard T1 */
-                        t2_timeout = client_compute_timeout(client, client->lease->lifetime, 0.5);
+                        client->t2_time = client_compute_timeout(client, client->lease->lifetime, 0.5);
                         client->lease->t2 = client->lease->lifetime / 2;
                 }
         } else {
                 /* fall back to the default timeouts */
-                t1_timeout = client_compute_timeout(client, client->lease->lifetime, 0.5);
+                client->t1_time = client_compute_timeout(client, client->lease->lifetime, 0.5);
                 client->lease->t1 = client->lease->lifetime / 2;
-                t2_timeout = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
+                client->t2_time = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
                 client->lease->t2 = (client->lease->lifetime * 7) / 8;
         }
 
         /* arm lifetime timeout */
         r = event_reset_time(client->event, &client->timeout_expire,
                              clock_boottime_or_monotonic(),
-                             lifetime_timeout, 10 * USEC_PER_MSEC,
+                             client->expire_time, 10 * USEC_PER_MSEC,
                              client_timeout_expire, client,
                              client->event_priority, "dhcp4-lifetime", true);
         if (r < 0)
                 return r;
 
         log_dhcp_client(client, "lease expires in %s",
-                        format_timespan(time_string, FORMAT_TIMESPAN_MAX, lifetime_timeout - time_now, USEC_PER_SEC));
+                        format_timespan(time_string, FORMAT_TIMESPAN_MAX, client->expire_time - time_now, USEC_PER_SEC));
```

Comment:
The log should be moved to the `if` condition below, or need to be adjusted, as the time difference may be negative.

Reply:
ok, i added a commit at the start of the series to fix this